### PR TITLE
Fix curly brace and interpolated string

### PR DESCRIPTION
**Pull request by bernd5**
fixes: #57750

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@bernd5 We've discussed this and are going to take this change.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:462]: i don't generally think this comment is that helpful.  But i also don't think it hurts too much to have.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:1182]: can we have an explicit test like these of hte case that regressed (not a parsing test, but a full compilation test).  we should also probably have the test validate the output received on multiple different framework versions.  Compiler can likely walk you through how to write those tests.

**Comment by jcouv** [comment]
We probably need to document a breaking change in https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%207.md

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs:3981]: can you add IDE tests in SemanticClassifierTests.cs (next to TestStringEscape) to validate the IDE side of this?  Thanks!

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler For eyes on this.  LDM approved us changing the behavior here as this is the behavior the runtime already changed to in .net core3.0.  This is now an impact when we emit interpolated string handlers.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:1182]: Also need an explicit test that shows we are generating the interpolated string handler code properly.

**Comment by CyrusNajmabadi** [comment]
@jcouv Agreed.  Does that go in with the PR, or is it a separate process?

**Comment by 333fred** [comment]
> @jcouv Agreed. Does that go in with the PR, or is it a separate process?

With this PR

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs
Comment (bernd5) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs:3981]: Oh, for that I think I need some help.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
Comment (bernd5) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:1280]: I created a test-theory. But I was quite a bit surprised that the test did not fail for Mscorlib40. Is the reason that the IL is identical and that the code is not executed under the target framework but the current one?

**Review by 333fred — DISMISSED** [review]
Compiler changes LGTM (commit 13)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs
Comment (bernd5) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs:3981]: @CyrusNajmabadi can you help me with such a test?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs:3981]: Yup. I can add this to this pr

**Comment by bernd5** [comment]
Is here still something to do?

**Comment by CyrusNajmabadi** [comment]
Yes.  We still need IDE tests.  I'm currently busy.  If you're able to add them yourself, that would be great.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:490]: i actually don't think this goes far enough.  If i look at the runtime impl, they just fail on usign a `{` in a format string as well.  I think we should follow that logic ourselves in the code that appears above this on line 474.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:490]: You can see the runtime logic here that errors: https://github.com/dotnet/runtime/blob/c72678bac90f9692ccea2787e22d88a6ae449cda/src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs#L1706

**Review by CyrusNajmabadi — CHANGES_REQUESTED** [review]
Changing to request changes.  I think we should match runtime here for both { and }.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:490]: this also matches teh grammar:

```
interpolation_format_character
    : '<Any character except \" (U+0022), : (U+003A), { (U+007B) and } (U+007D)>'
```

Comment (bernd5) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:489]: An error is now emitted (and I add tests for those)

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:71]:
```suggestion
5. Starting with Visual Studio 17.1, format specifiers in interpolated strings can not contain curly braces (either `{` or `}`). In previous versions `{{` was interpreted as an escaped `{` and `}}` was interpreted as an escaped `}` char in the format specifier. Now the first `}` char in a format specifier ends the interpolation, and any `{` char is an error.
```

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:480]: why was this change made?  it doesn't actually seem to match anything.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:486]: why was this change made?

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:495]: why the `:` part?

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:71]: @333fred do we specify a VS version?  or a compiler/lang version?

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:480]: i think we should make the minimal change necessary to match the runtime.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs:3981]: IDE tests added.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (bernd5) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:495]: Because the colon is not allowed according to the spec

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md
Comment (bernd5) [reply docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:71]: AFAIK the old behaviour is a spec violation which we do not want to preserve - so I think the VS version is fine

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:495]: i don't think we should be making any unnecessary changes.  esp. those that might break people.  The only approved change we ran by the LDM was updating curly brace matching to match what the runtime does. 

Unless the runtime also has this behavior for things like `:` and `"`, then we should not touch it.

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md
Comment (CyrusNajmabadi) [reply docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:71]: I'm not sure the relation though.  This compiler change will happen regardless of VS version.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (bernd5) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:495]: Okay, I'm fine to allow those - for me even '{' would be okay

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (bernd5) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:486]: the same as for ':' - '"' is not allowed
But we could ignore the spec here, too.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (bernd5) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:480]: it matches for \u0022 and \u003A

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:495]: we should disallow `{` so that we can be in sync with teh runtime.  The goal here is to get us operating the same way so we're in agreemnt on the processing.  ideally we shoudl only make one change and be done here :)

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md
Comment (333fred) [reply docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:70]: We specify VS version.

File: src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs:457]: Unlike Cyrus, I think this addendum is actively harmful. It reads like something we're doing currently. I think just the spec part is fine and won't lead to future confusion.

**Comment by 333fred** [comment]
Hopefully final question is: do we need to make a change in VB parsing as well?

**Comment by CyrusNajmabadi** [comment]
My feeling is that vb is likely fine. Since it doesn't support interpolated string handlers right?  So it will still emit string. Format and will be ok.

**Comment by 333fred** [comment]
> My feeling is that vb is likely fine. Since it doesn't support interpolated string handlers right? So it will still emit string. Format and will be ok.

It doesn't, but it will still parse differently, right? If we're ok with that, then ok. Just wanted a check on whether we think that parsing difference is ok.

**Comment by 333fred** [comment]
@bernd5 looks like there's a couple of nullable issues:

>>> D:\a\_work\1\s\src\EditorFeatures\CSharpTest\Classification\SemanticClassifierTests.cs(2609,31): error CS8602: Dereference of a possibly null reference. [D:\a\_work\1\s\src\EditorFeatures\CSharpTest\Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests.csproj]
D:\a\_work\1\s\src\EditorFeatures\CSharpTest\Classification\SemanticClassifierTests.cs(2623,32): error CS8600: Converting null literal or possible null value to non-nullable type. [D:\a\_work\1\s\src\EditorFeatures\CSharpTest\Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests.csproj]

**Comment by bernd5** [comment]
Yes Cyrus removed #nullable disable

**Comment by CyrusNajmabadi** [comment]
Sorry, will fix.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by 333fred** [comment]
> It doesn't, but it will still parse differently, right? If we're ok with that, then ok. Just wanted a check on whether we think that parsing difference is ok.

Before I sign off I'd like @AlekseyTs's thoughts on this as well. Note that this won't change _runtime_ semantics, only the parse tree.

**Comment by bernd5** [comment]
In my eyes VB should be changed, too. But in a separate PR.

**Comment by 333fred** [comment]
I spoke with @AlekseyTs offline about this. We're fine with not changing VB in this PR, and I filed https://github.com/dotnet/roslyn/issues/58043 as a help-wanted issue if someone wants to correct VB parsing in the future.

**Review by 333fred — APPROVED** [review]
LGTM (commit 33). @dotnet/roslyn-compiler for a second review.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson can you look at this community PR as well?  It will impact the raw-string-literal work as well, so it would be good to have your eyes on it.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @chsienki can you ptal here?  thanks!

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
Comment (chsienki) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:437]: Nit: should this be English to be consistent? 

(Although I quite like the idea of having different languages in the code base 😊)

**Review by chsienki — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
THanks @bernd5 !  I'll make sure the raw-string-literals work picks this up as well.

**Pull request merged**
